kernel/inductive.ml

```diff
@@ -330,33 +330,44 @@ let check_allowed_sort ksort specif =
     let s = inductive_sort_family (snd specif) in
     raise (LocalArity (Some(elim_sort specif, ksort,s,error_elim_explain ksort s)))
 
-let is_correct_arity env c pj ind specif params =
+let check_correct_arity env c pj ind specif params =
   let arsign,_ = get_instantiated_arity ind specif params in
-  let rec srec env pt ar =
+  let rec srec env ar pt =
     let pt' = whd_all env pt in
-    match kind pt', ar with
-      | Prod (na1,a1,t), (LocalAssum (_,a1'))::ar' ->
-          let () =
-            try conv env a1 a1'
-            with NotConvertible -> raise (LocalArity None) in
-          srec (push_rel (LocalAssum (na1,a1)) env) t ar'
-      (* The last Prod domain is the type of the scrutinee *)
-      | Prod (na1,a1,a2), [] -> (* whnf of t was not needed here! *)
-         let env' = push_rel (LocalAssum (na1,a1)) env in
-         let ksort = match kind (whd_all env' a2) with
-         | Sort s -> Sorts.family s
-         | _ -> raise (LocalArity None) in
-         let dep_ind = build_dependent_inductive ind specif params in
-         let _ =
-           try conv env a1 dep_ind
-           with NotConvertible -> raise (LocalArity None) in
-           check_allowed_sort ksort specif
-      | _, (LocalDef _ as d)::ar' ->
-          srec (push_rel d env) (lift 1 pt') ar'
-      | _ ->
-          raise (LocalArity None)
+    match ar, kind pt' with
+    | (LocalAssum (_,a1))::ar', Prod (na1,a1',t) ->
+      let () =
+        try conv_leq env a1 a1'
+        with NotConvertible -> raise (LocalArity None) in
+      srec (push_rel (LocalAssum (na1,a1)) env) ar' t
+    (* The last Prod domain is the type of the scrutinee *)
+    | [], Prod (na1,a1',a2) ->
+      let env' = push_rel (LocalAssum (na1,a1')) env in
+      let ksort = match kind (whd_all env' a2) with
+      | Sort s -> Sorts.family s
+      | _ -> raise (LocalArity None)
+      in
+      let dep_ind = build_dependent_inductive ind specif params in
+      let () =
+        (* This ensures that the type of the scrutinee is <= the
+           inductive type declared in the predicate. *)
+        try conv_leq env dep_ind a1'
+        with NotConvertible -> raise (LocalArity None)
+      in
+      let () = check_allowed_sort ksort specif in
+      (* We return the "higher" inductive universe instance from the predicate,
+         the branches must be typeable using these universes. *)
+      let (pind, _args) =
+        try find_rectype env a1'
+        with Not_found ->
+          anomaly (Pp.str "Not an inductive type but convertible to an inductive type")
```

Comment:
I don't think we need to catch here

Reply:
Well, I didn't know which was preferable, we know that a1' must be an inductive given the previous cumulativity check.

Reply:
I removed it but added an explanatory comment.